_drafts/implementing-state-machines-in-postgresql.md

```diff
@@ -0,0 +1,301 @@
+---
+layout: post
+title: "Implementing State Machines in PostgreSQL"
+---
+
+[Finite-state machines](https://en.wikipedia.org/wiki/Finite-state_machine)
+(FSMs) are a wonderful model of computation that has many practical
+applications. One of my favorites is turning business logic into simple FSMs.
+For example consider the following requirements for an order management system:
+
+* Orders must not ship before they are paid.
+* Orders can be canceled, but only if they haven't shipped yet.
+* An order that has already been paid, needs to be refunded after canceling.
+
+Informally, turning such a list of requirements into an FSM is simply the
+process of coming up with: a set of states an order can be in, a set of events,
+a transition function that maps the combination of each state and event to a
+new state, and an initial state.
+
+In practice this is usually done by drawing up a directed graph that shows how
+the different states are connected to each other via events:
+
+<img src="{% asset_path orders.svg %}" alt="Finite-state machine for an order management system in PostgreSQL" />
+
+Besides guiding our implementation, these graphs can be very useful for
+discussions and analysis. For example you can immediately see how there is no
+way for a customer to return an order after it has shipped yet. Additionally
+the process of naming the involved states and events automatically creates a
+precise language that can be adopted by all stakeholders.
+
+If you paid close attention so far, you might have noticed that the graph above
+does not show the combination of all states and events which simply implies
+that the missing combinations lead to an error state. The [complete FSM]({%
```

Comment:
I'd change this into this:

"You might have noticed that the graph above does not show all possible combinations of all states and all events. The reason is that the missing combinations lead to error states."